src/celeritas/optical/CerenkovPreGenerator.hh

```diff
@@ -0,0 +1,154 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2024 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/optical/CerenkovPreGenerator.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "corecel/Assert.hh"
+#include "corecel/Macros.hh"
+#include "celeritas/phys/ParticleTrackView.hh"
+#include "celeritas/random/distribution/PoissonDistribution.hh"
+#include "celeritas/track/SimTrackView.hh"
+
+#include "CerenkovData.hh"
+#include "CerenkovDndxCalculator.hh"
+#include "OpticalDistributionData.hh"
+#include "OpticalPropertyData.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+/*!
+ * Sample the number of Cerenkov photons to be generated by
+ * \c CerenkovGenerator and populate \c OpticalDistributionData values using
+ * Param and State data.
+ * \code
+    CerenkovPreGenerator::OpticalPreGenStepData step_data;
+    // Populate step_data
+
+   CerenkovPreGenerator pre_generate(particle_view,
+                                     sim_view,
+                                     material,
+                                     properties->host_ref(),
+                                     params->host_ref(),
+                                     step_data);
+
+    auto optical_dist_data = pre_generate(rng);
+    if (optical_dist_data)
+    {
+        CerenkovGenerator cerenkov_generate(... , optical_dist_data, ...);
+        cerenkov_generate(rng);
+    }
+ * \endcode
+ */
+class CerenkovPreGenerator
+{
+  public:
+    // Placeholder for data that is not available through Views
+    struct OpticalPreGenStepData
+    {
+        real_type time{};  //!< Pre-step time
+        EnumArray<StepPoint, OpticalStepData> points;  //!< Pre- and post-steps
```

Comment:
We may end up moving this outside this class since the scintillation pre-generator will use it too. We also might be able to get away with just storing the pre-step speed and position and getting the post-step values from the particle/geo track views (don't have to do that now, just something to consider in the future).